### PR TITLE
[WIP] Add `chainerx.hinge`

### PR DESCRIPTION
**Pull request by aksub99**
Add `chainerx.hinge`

### CODE REVIEW

**Comment by aksub99** [comment]
In `routines/loss.cc`, I have used `At` to access a particular element. But, it seems to be hidden in `indexing.h`. From what I understand, there are 3 possible options:
1) expose `At` and use already defined methods
2) implement a custom kernel for `hinge`
3) Alternative implementation using defined methods without the necessity of `At`.(If so, I would appreciate some suggestions for the same)
What could be the best option here?

File: chainerx_cc/chainerx/routines/loss.cc [hunk 1]
```diff
@@ -24,4 +28,19 @@ Array HuberLoss(const Array& x1, const Array& x2, Scalar delta) {
     return Where(abs_a < delta_array, 0.5 * Square(a), delta * (abs_a - Scalar{0.5} * delta));
 }
 
+Array Hinge(const Array& x1, const Array& x2, const std::string& norm) {
+    int64_t num = x1.shape()[0];
+    Array bottom_diff = Copy(x1);
+    std::vector<Array> indices = {Arange(num), x2};
+    Array concat = Transpose(Concatenate(indices, 0));
+    IMultiply(At(bottom_diff, concat), -1);
```
Comment (kshitij12345) [inline chainerx_cc/chainerx/routines/loss.cc]: You can turn this into.
```CXX
Array selected = bottom_diff.At(concat);
selected *= -1;
```
As they are methods for `Array`. This should compile.

Comment (aksub99) [reply chainerx_cc/chainerx/routines/loss.cc]: Thanks @kshitij12345 .The only problem seems to be that `At` expects a `vector` with `ArrayIndex` as element type.

Comment (kshitij12345) [reply chainerx_cc/chainerx/routines/loss.cc]: I guess, you can create `ArrayIndex` using `Slice` object with your range of indices.

File: chainerx_cc/chainerx/routines/loss.cc [hunk 2]
```diff
@@ -24,4 +28,19 @@ Array HuberLoss(const Array& x1, const Array& x2, Scalar delta) {
     return Where(abs_a < delta_array, 0.5 * Square(a), delta * (abs_a - Scalar{0.5} * delta));
 }
 
+Array Hinge(const Array& x1, const Array& x2, const std::string& norm) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/loss.cc]: Please raise `DimensionError` when `ndim` and `shape` of `x1` or `x2` are wrong.

File: chainerx_cc/chainerx/routines/loss.cc (hunk 1 above)
Comment (asi1024) [reply chainerx_cc/chainerx/routines/loss.cc]: `IMultiply` cannot be used for an array required grad.

Comment (asi1024) [reply chainerx_cc/chainerx/routines/loss.cc]: I think there is no need to use `At`.
How about implementing like as follows for example?
```py
Array one_minus_diff = Where(ExpandDims(x2, 1) == Arange(num, x2.dtype), 0, 2);
```

**Comment by aksub99** [comment]
I tried out what you have suggested, but it turns out that the dimensions of `one_minus_diff` do not match that of `1 + self.bottom_diff` in `Chainer`'s implementation.

File: chainerx_cc/chainerx/routines/loss.cc (hunk 2 above)
Comment (asi1024) [reply chainerx_cc/chainerx/routines/loss.cc]: `norm` should not be passed as a string value. How about
1. defining an enum class or
2. passing as a float value?

Comment (aksub99) [reply chainerx_cc/chainerx/routines/loss.cc]: Could I know why `string` is not preferred?
If defined as `float`, what could be the values of `norm` corresponding to `L1` and `L2` ?

Comment (asi1024) [reply chainerx_cc/chainerx/routines/loss.cc]: `string` is not preferred because of the performance.

Comment (aksub99) [reply chainerx_cc/chainerx/routines/loss.cc]: > Could I know why `string` is not preferred?
> If defined as `float`, what could be the values of `norm` corresponding to `L1` and `L2` ?

Thanks.
Could you please also answer the second question?

Comment (asi1024) [reply chainerx_cc/chainerx/routines/loss.cc]: I mean, let's support general L-p norm. In this case, `float norm` represents the p.

Comment (aksub99) [reply chainerx_cc/chainerx/routines/loss.cc]: Oh, I get it now. So, why not define it as an `int` since 1 and 2 are integer values?

Comment (aksub99) [reply chainerx_cc/chainerx/routines/loss.cc]: Do you mean to say that we can allow `p` value to be non integers as well to make it general?

Comment (asi1024) [reply chainerx_cc/chainerx/routines/loss.cc]: Yes. L-p norm is defined for all real p >= 1.

**Pull request closed without merging**